## Code for computing summary statistics for JIA subtypes.

## parse options first in case there is a problem otherwise slow.
library(optparse)
option_list = list(
        make_option(c("-o", "--out"), type="character", default=NULL,
              help="output dir", metavar="character"),
        make_option(c("-s", "--sample_file"), type="character", default='cc',
              help="list of samples to include", metavar="character"),
        make_option(c("-i", "--support_file"), type="character", default=NULL,
              help="Support file generated by create_t1d_support.R", metavar="character")
        )


## valid phenotypes


opt_parser = OptionParser(option_list=option_list);
args = parse_args(opt_parser)
if (is.null(args$out) | is.null(args$sample_file) | is.null(args$support_file)){
	print_help(opt_parser)
	stop("At least one argument must be supplied (output file)", call.=FALSE)
}

## phenotypes that are available are

#args<-list(
#  input='/scratch/wallace/JIA-2017-data/annotsnpstats-22.RData',
#  out='/scratch/wallace/JIA-2017-data/summary_stats_OB/',
#  phenotype='sys'
#  )

## for chromosome 1 the largest there are 562166 snps it takes approx 5 seconds to fit glm for 100

library(data.table)
library(magrittr)
library(rmeta)
library(SNPlocs.Hsapiens.dbSNP144.GRCh37)
library(snpStats)
## get the data
#args<-list(input='/home/ob219/scratch/as_basis/t1d_gwas_sim/support/22_support.RData',
  #sample_file='/home/ob219/scratch/as_basis/t1d_gwas_sim/sample_lists/cases100control100_3')

if(FALSE){
  ## code here for sampling individuals for size t1d_titrations
  ## NOTE THIS IS NOT RUN
  load(args$support_file)
  out.dir<-'/home/ob219/scratch/as_basis/t1d_gwas_sim/sample_lists'
  wtccc.support<-data.table(wtccc.support)
  wtcc.support[,cohort='WTCCC']
  t1dgc.support<-data.table(t1dgc.support)
  t1dgc.support[,cohort='T1DGC']
  all.support<-rbind(wtccc.support,t1dgc.support)
  cases<-all.support[which(all.support$t1d==2),]$id
  controls<-all.support[which(all.support$t1d==1),]$id
  sizes<-c(50,100,200,800,1600,2500,3000,4000)
  n<-10

  for(i in sizes){
    for(j in 1:n){
      write(c(sample(cases,i),controls),file=file.path(out.dir,sprintf("cases%dcontrol%d_%d",i,length(controls),j)))
    }
  }
}

message(sprintf("Getting data from %s",args$support_file))

## trim the support and snp matrix
load(args$support_file)

## next we need to trim these so just samples that we have in sample file

samples<-scan(args$sample_file,"character()")
wtccc.keep<-which(wtccc.support$id %in% samples)
t1dgc.keep<-which(t1dgc.support$id %in% samples)

t1dgc.support$cc <- factor(t1dgc.support$t1d - 1)
wtccc.support$cc <- factor(wtccc.support$t1d - 1)
message(sprintf("Fitting models for snps"))
illumina.6 <- snp.rhs.estimates(cc ~ 1, snp.data=t1dgc.snps[t1dgc.keep,],data=t1dgc.support[t1dgc.keep,])
affy.6 <- snp.rhs.estimates(cc ~ 1, snp.data=wtccc.snps[wtccc.keep,],data=wtccc.support[wtccc.keep,])
illumina.6i <- snp.rhs.estimates(cc ~ 1, snp.data=t1dgc.snps[t1dgc.keep,],data=t1dgc.support[t1dgc.keep,],rules=i2a)
affy.6i <- snp.rhs.estimates(cc ~ 1, snp.data=wtccc.snps[wtccc.keep,],data=wtccc.support[wtccc.keep,],rules=a2i)

message("Reformatting and computing summary statistics")

myunlist <- function(k) {
       nulls <- sapply(k,is.null)
       if(any(nulls)) {
           for(i in which(nulls))
               k[[i]] <- NA
       }
       unlist(k)
}

extract <- function(x) {
     data.frame(row.names=lapply(x,"[[","beta") %>% names(),
                snp=lapply(x,"[[","beta") %>% names(),
                beta=lapply(x,"[[","beta") %>% myunlist(),
                v.beta=lapply(x,"[[","Var.beta") %>% myunlist())
}
i6 <- rbind(extract(illumina.6),extract(illumina.6i))
a6 <- rbind(extract(affy.6),extract(affy.6i))

tmp <- merge(i6,a6,by="snp",suffixes=c(".i",".a"))
tmp$beta.meta <- tmp$v.beta.meta <- NA
tmp <- merge(tmp,ss[,c("b36_chr","b36_start","affy_id","wtccc.affy_alleles","affy_strand",
                          "ilmn_id","ilmn_alleles","ilmn_strand","i.switched")],
                by.x="snp",by.y=0,all.x=TRUE)

for(i in 1:nrow(tmp)) {
      if(is.na(tmp$beta.i[i])) {
          tmp$beta.meta[i] <- tmp$beta.a[i]
          tmp$v.beta.meta[i] <- tmp$v.beta.a[i]
      } else if(is.na(tmp$beta.a[i])) {
          tmp$beta.meta[i] <- tmp$beta.i[i]
          tmp$v.beta.meta[i] <- tmp$v.beta.i[i]
      } else {
          ms <- meta.summaries(d=c(tmp$beta.i[i],tmp$beta.a[i]),
                               se=sqrt(c(tmp$v.beta.i[i],tmp$v.beta.a[i])))
          tmp$beta.meta[i] <- ms$summary
          tmp$v.beta.meta[i] <- ms$se.summary^2
      }
}

tmp$p <- pnorm(abs(tmp$beta.meta/sqrt(tmp$v.beta.meta)),lower.tail=FALSE)*2

tmp$consensus.alleles <- tmp$wtccc.affy_alleles
tmp$consensus.alleles[tmp$wtccc.affy_alleles==""] <- tmp$ilmn_alleles[tmp$wtccc.affy_alleles==""]
output<-data.table(tmp)[,.(snp,b36_chr,b36_start,consensus.alleles,beta.meta,p)]

## convert to build 37
snps <- SNPlocs.Hsapiens.dbSNP144.GRCh37
chr<-unique(output$b36_chr)
f22<-data.table(as.data.frame(snpsBySeqname(snps,sprintf("ch%s",chr))))
setkey(f22,RefSNP_id)
setkey(output,snp)

output<-f22[output][,.(RefSNP_id,b36_chr,pos,consensus.alleles,exp(beta.meta),p)]
output<-output[,c('a1','a2'):=tstrsplit(consensus.alleles,'/')][order(pos),.(RefSNP_id,b36_chr,pos,a1,a2,V5,p)] %>% setnames(.,c('id','chr','position','a1','a2','or','p.val'))

out.file<-sprintf("%s_%s.RDS",basename(args$sample_file),chr)
out.file<-file.path(args$out,out.file)
saveRDS(output,file=out.file)
message(sprintf("Wrote file to %s",out.file))

if(FALSE){
  ## code for assembling
  out.dir<-'/home/ob219/scratch/as_basis/t1d_gwas_sim/input_files/size_titration/'
  support.dir<-'/scratch/ob219/as_basis/support_tab'
  ref_af_file<-file.path(support.dir,'as_basis_snps_with_alleles.tab')
  bsnps <- fread(ref_af_file)
  bsnps[,pid:=paste(chr,position,sep=':')]
  setkey(bsnps,pid)
  DT<-data.table(list.files(path='.',pattern='*.RDS'))
  DT[,c('ss','sim','chr'):=tstrsplit(V1,'_')]
  SDT<-DT[,list(N=.N),by=c('ss','sim')]
  for(i in 1:nrow(SDT)){
    mss<-SDT[i,]$ss
    simno<-SDT[i,]$sim
    ofile<-sprintf("%s_%s.tab",mss,simno)
    message(sprintf("Processing %s",ofile))
    all<-rbindlist(lapply(subset(DT,ss==mss & sim==simno)$V1,readRDS))
    all[,pid:=paste(chr,position,sep=':')]
    ## need to check that all basis SNPs are available
    setkey(all,pid)
    res<-all[bsnps]
    if(nrow(res)!=nrow(bsnps)){
      stop(sprintf("Number of SNPs are missing got %d had %d",nrow(res),nrow(bsnps)))
    }
    res<-res[order(chr,position),.(id,chr,position,p.val,or,a1,a2)]
    write.table(res,file=file.path(out.dir,ofile),sep="\t",row.names=FALSE,quote=FALSE)
  }
}
